evap/staff/views.py

```diff
@@ -790,12 +798,61 @@ def questionnaire_copy(request, questionnaire_id):
         else:
             return render(request, "staff_questionnaire_form.html", dict(form=form, formset=formset))
     else:
-        questionnaire = get_object_or_404(Questionnaire, id=questionnaire_id)
+        form, formset = get_identical_form_and_formset(questionnaire_id)
+        return render(request, "staff_questionnaire_form.html", dict(form=form, formset=formset))
+
+
+def find_questionnaire_name(old_questionnaire):
+    timestamp = datetime.date.today()
+    new_name_de = old_questionnaire.name_de + ' pre ' + str(timestamp)
+    new_name_en = old_questionnaire.name_en + ' pre ' + str(timestamp)
+
+    if Questionnaire.objects.filter(Q(name_de=new_name_de) | Q(name_en=new_name_en)):
+        version = 1
+        while Questionnaire.objects.filter(Q(name_de=new_name_de + ' ({})'.format(version)) |
+                                           Q(name_en=new_name_en + ' ({})'.format(version))):
+            version += 1
+        new_name_de += ' ({})'.format(version)
+        new_name_en += ' ({})'.format(version)
+    return new_name_de, new_name_en
+
+
+@staff_required
+def questionnaire_new_version(request, questionnaire_id):
+    if request.method == "POST":
+        questionnaire = Questionnaire()
         InlineQuestionFormset = inlineformset_factory(Questionnaire, Question, formset=AtLeastOneFormSet, form=QuestionForm, extra=1, exclude=('questionnaire',))
 
-        form = QuestionnaireForm(instance=questionnaire)
-        formset = InlineQuestionFormset(instance=questionnaire, queryset=questionnaire.question_set.all())
+        form = QuestionnaireForm(request.POST, instance=questionnaire)
+        formset = InlineQuestionFormset(request.POST.copy(), instance=questionnaire, save_as_new=True)
+
+        old_questionnaire = get_object_or_404(Questionnaire, id=questionnaire_id)
+
+        # Find new title for old Questionnaire.
+        new_name_de, new_name_en = find_questionnaire_name(old_questionnaire)
+
+        old_name_de = old_questionnaire.name_de
+        old_name_en = old_questionnaire.name_en
 
+        # Change old name before checking Form.
+        old_questionnaire.name_de = new_name_de
+        old_questionnaire.name_en = new_name_en
+        old_questionnaire.save()
+
+        if form.is_valid() and formset.is_valid():
+            form.save()
+            formset.save()
+
+            messages.success(request, _("Successfully created questionnaire."))
+            return redirect('staff:questionnaire_index')
+        else:
+            # Revert old Questionnaire to old title.
+            old_questionnaire.name_de = old_name_de
+            old_questionnaire.name_en = old_name_en
+            old_questionnaire.save()
```

Comment:
if something crashes before this is done, the database is changed but shouldn't be. so please find a way to do the validation without changing the DB or use transactions or something like that.